core/messaging/TFMessaging.py

```diff
@@ -0,0 +1,234 @@
+
+    def non_nullable_constraint_failure_message(self, validation_failure) -> tuple[str, str, str, str]:
+        """Generate error message components for NonNullableConstraintFailure.
+
+        This function returns error message components in the case of a NonNullableConstraintFailure.
+        In instances where the Unit of Measurement is null, a distinct error message is necessary.
+        The function distinguishes between Outputs and Outcomes and adjusts the error message accordingly.
+
+        return: tuple[str, str, str]: A tuple containing the sheet name, section, and error message.
+        """
+        sheet = INTERNAL_TABLE_TO_FORM_SHEET[validation_failure.table]
+        column, section = INTERNAL_COLUMN_TO_FORM_COLUMN_AND_SECTION[validation_failure.column]
+
+        cell_index = construct_cell_index(
+            table=validation_failure.table, column=validation_failure.column, row_index=validation_failure.row_index
+        )
+
+        message = self.msgs.BLANK
+        if sheet == "Project Outputs":
+            if column == "Unit of Measurement":
+                message = self.msgs.BLANK_UNIT_OF_MEASUREMENT
+            if column == "Financial Year 2022/21 - Financial Year 2025/26":
+                message = self.msgs.BLANK_ZERO
+        elif sheet == "Outcomes":
+            if column == "Unit of Measurement":
+                message = self.msgs.BLANK_UNIT_OF_MEASUREMENT
+            if column == "Financial Year 2022/21 - Financial Year 2025/26":
+                section = "Outcome Indicators (excluding footfall) / Footfall Indicator"
+                message = self.msgs.BLANK_ZERO
+                cell_index = get_cell_indexes_for_outcomes(validation_failure.failed_row)
+        elif sheet == "Funding Profiles":
+            message = self.msgs.BLANK_ZERO
+        elif section == "Programme-Wide Progress Summary":
+            message = self.msgs.BLANK
+
+        return sheet, section, cell_index, message
+
+    def towns_fund_round_four_validation_failure(self, validation_failure) -> tuple[str, str, str, str]:
+        cell_index = construct_cell_index(
+            validation_failure.sheet, validation_failure.column, validation_failure.row_index
+        )
+        sheet = INTERNAL_TABLE_TO_FORM_SHEET[validation_failure.sheet]
+        return sheet, validation_failure.section, cell_index, validation_failure.message
+
+    def wrong_input_failure_message(self, validation_failure) -> tuple[str | None, str | None, str]:
+        return None, None, self.tf_msgs.PRE_TRANSFORMATION_MESSAGES[validation_failure.value_descriptor]
+
+    def unauthorised_submission_failure(self, validation_failure) -> tuple[str | None, str | None, str]:
+        places_or_funds = join_as_string(validation_failure.expected_values)
+        message = self.msgs.UNAUTHORISED.format(
+            wrong_place_or_fund_type=validation_failure.entered_value, allowed_places_or_fund_types=places_or_funds
+        )
+        return None, None, message
```

Comment:
No, because these are pre-transformation messages they have a different function signature. This migth lead to confusion as there is no clear way to tell what is a pre and post validation message at this level